### DISCUSSION TITLE
Add a spell checker as "extension".

### DISCUSSION THREAD

**Discussion by GilShoshan94** [rust-lang/rust-analyzer#14389] [Ideas]
Hi,

Now, we have a code _**formater**_ with `rustfmt`, we have a _**linter**_ with `cargo check` or `cargo clippy`, why not a _**spell checker**_?

A _**spell checker**_ sounds like an easy addition, it's basically just a linter specialize in spelling.

I know a spell checker sounds a bit out of scope for Rust Analyzer, but is it really ? I mean we do have a formater already, having code well formatted help with consistency and to focus on the logic instead of the aesthetic, I would argue that the same holds for spelling.

[**Typos**](https://github.com/crate-ci/typos) is a source code spell checker written in Rust, it supports [several interfaces](https://github.com/crate-ci/typos#custom), file, `stdin` to `stdout` and `--format json`. 
Its design choice is described [here](https://github.com/crate-ci/typos/blob/master/docs/design.md).

Instead of having Typos to make it's own LSP (if it's even relevant, as I heard that editors will not work with multiple LSP servers for one language, but could not confirm), or having several custom plugging for each IDE, it could instead work like a "linter supplier" to existing LSP, such as Rust Analyzer to start with.

What would it requires from Typos to make it work with Rust Analyzer and is it simple enough/possible for Rust Analyzer to add a spell checker source too ?

(Also I hope this is the right place to open this conversation, if needed, I can open an issue "feature request" instead)
(Mention to https://github.com/crate-ci/typos/issues/263 and @epage)

**Comment by Veykril** [comment]
A spell checker is vastly different from a code formatter. A code formatter needs knowledge about the structure of the language it targets, a spell checker does not. A spell checker can target any source files from any language without requiring any specific support for languages while a code formatter is usually tied to a certain language.

As such this is very much out of scope for rust-analyzer and would be better served as its own extension

**Reply by GilShoshan94** [reply]
Hi, 

Thank you for responding.

The way r-a runs rustfmt is not just by passing the file to it ?

In theory, how hard would it be for r-a to pass the input to the spell checker, and get back a json formatted reponse (same format than the linter) and use that to add more underlines, warnings, fix suggestion, ... back in the IDE ?

Otherwise, can you advise if the right way for the spell checker is to implement lsp or to make a normal extension ?

**Reply by Veykril** [reply]
Whether r-a formats code by passing it to rustfmt or not isn't relevant here. rustfmt is a rust specific tool (that is also shipped with the rust toolchain) and as such support for it has a place in a rust IDE. Note that for running `cargo check` from within r-a, we are actually already contemplating whether it would make more sense to split that out from the server or not as well.

Using a language server (that is LSP) for a spell checker makes little sense, a spell checker is not a tool that provides intellisense or similar functionality for a given language. A simple extension whose sole purposes is to spell check files makes more sense.

**Reply by GilShoshan94** [reply]
Thank you for the insight !

I understand, an LSP for a spell checker is overkill, the only reason it was mention was for the standardization between all the IDE implementing the lsp protocol.

 Are you aware of another solution than writing a custom extension for each IDE, that would work for all ?

**Comment by Veykril** [comment]
No I am not

**Reply by GilShoshan94** [reply]
Too bad... it would be cool if there was a standard such as LSP but for tools such as linters, formatter and spell checker, that only need to pass on the file and give a feedback to IDE, without all the complex knowledge about the structure of the language of the code.

Thank you for your time an for sharing your knowledge.
I am closing this discussion now.

Have a nice day.

**Comment by GilShoshan94** [comment]
Hi @Veykril,
Sorry to reopen...

I researched about the subject a bit more, and I am still not totally conviced that this is really out of scope...

@matklad seems to think that an ideal spell checker for code should be aware of the language (comment [here ](https://github.com/microsoft/vscode/issues/20266#issuecomment-470620828) and [here](https://github.com/rust-lang/rust-analyzer/pull/332#discussion_r243914251)).

Also looking more into Typos, it seems to be [language aware at some extend](https://github.com/crate-ci/typos/blob/master/docs/comparison.md) (see `typos --type-list`). @epage can talk more about this.

They already discuss the idea [to integrate LSP and spellcheckers](https://www.reddit.com/r/rust/comments/nlh4u6/comment/gzqgsoi/?utm_source=share&utm_medium=web2x&context=3).

Typos is a mature tool, past 1.0, has many features, differenciates between identifiers and words, and support several outputs formats, including json.

I really feel that it can improve the devlopper experience and it really seems that it would not even be that hard to plug this into r-a, all what the spell checker needs, is to be fed, which r-a does already for rustfmt, cargo-check and cargo-clippy, and the outputs needs to be understood by r-a to know what to underlines in the IDE and provide the fix suggestions in the IDE, exactly like with the linter. So if r-a has already a specific format that it expect from the linter, why not simply use the same for Typos ?

I would like to hear what @matklad think of this.

Please think of it for a couple of days. I really think it would be a great feature that cost near to nothing to r-a and would improve dev experience.

After I will close this I promess.

**Reply by Veykril** [reply]
No need to apologize for re-opening. I haven't looked through all the things you've linked but there are some valid arguments as to why a spell checker may be language aware in them that I didn't think of before. I'll come back to this with some more thoughts later on then, but maybe matklad will also have left some comments by then (as he does seem to have some ideas around the topic).

**Reply by GilShoshan94** [reply]
@matklad, can you comment on this please (whenever you can of course) ?

Also @epage, do you see any features you can get by beeing more integrated into an LSP and have more meta data about the code beeing check ?

Thinking more about this, if you know when a token is a function or a method call, you can skip the spell check as it cannot be changed there by the user, only spell check declarations and definitions.

Also in doc comments, we can also know if this is sample code with ` ```rust .... ``` ` blocks and with tokens that are surounded by ``.

It can reduce a lot of noise from the spell checker and be more performant I think.

**Reply by epage** [reply]
> Thinking more about this, if you know when a token is a function or a method call, you can skip the spell check as it cannot be changed there by the user, only spell check declarations and definitions.

This would be the biggest advantage.  typos does a lot of work to try to avoid spell checking non-words like hashes.  Granted, a lot of programs could be helped by spell checking user-facing text, so maybe those heuristics would still be needed.

rust-lang/rust-clippy#2508 would be an alternative to rust-analyzer as it too can look at definitions rather than uses, there is the `--fix` flag for corrections, and clippy has a setting to not lint exported APIs to avoid breaking changes.

**Comment by matklad** [comment]
Some thoughts:

* I'd be against using typos as it is:
  * not ubiquitous enough to justify it's inclusion on the grounds of practicality
  * doesn't have the right design for the problem at hand, to to justify inclusion on the grounds of "this is the beginning of a perfect solution"

That's the general design principle for features in rust-analyzer: 

* either the feature is implemented with technically optimal design (it doesn't matter if its incomplete or buggy, if the foundation is right. That's basically the whole rust-analyzer vs rustc thing)
* or we begrudgingly accept something which has wrong design, but ubiquitous (eg, we shell out `cargo check` or `rustfmt` for the time being while we haven't implemented the right thing ourselves yet)

For spell checker, a significant chunk of the logic indeed belongs to rust-analyzer. Specifically, rust-analyzer should: 

* separate keywords from identifiers
* separate _reference_ identifiers from _definition_ identifiers. Typos should be checked only for definitions. Its compiler's job to point typos in the references.
* separate declarations from definitions. Function names should be checked in traits, but not in trait impls.
* find comments
* unesape strings (_values_ of string literals should be checked for typos, not the source syntax)

I think adding the following API would make sense: 

```rust
struct SpellWord {
    text: String,
    range: TextRange
}

impl Analysis {
  /// Returns the list of words for spell-checking.
  /// Note that this returns strings and not just ranges, because the value of
  /// the word to spell-check might differ from the text at range due to escaping. 
  pub fn spell_words(&self, file_id: FileId) -> Vec<SpellWord>

  /// Correct spelling (this might trigger rename refactor).
  pub fn correct_spelling(&self, file_id: FileId, word: SpellWord, correction: String) -> SourceChange
}
```

Implementaiton wise, I think to be pedantically correct we need to run this on macro-expanded code, but I think it'd be more robust to restrict this feature to be based strictly on the syntax.

**Reply by matklad** [reply]
I am not sure if it makes sense to expose this API as actual LSP diagnostics, as we need two things for it:

* a spell-checker library (perhaps typos is that? not sure)
* a new bit of user-visible state to configure project-specific and user-specific dictionary 

The last bit is something which I would be particularly reluctant to add

**Reply by epage** [reply]
Instead of using `typos-cli` as a binary, you can use it as a library or build your own logic on top of `typos-dict` / `typos-vars`.  There hasn't been too much interest in it so far, so I've not done a tone of work polishing it up.

**Reply by GilShoshan94** [reply]
Thank you both @matklad and @epage  for taking part to the discussion !

From reading your exchanges, here is my inputs:

#### General:
- A user-visible state to configure project-specific and user-specific dictionary is inevitable in my opinion. I don't know how the format would look but it should support some options. More on this later.
- The spell checking should be opt-in.

#### Implementation:
After reading your drafted API `SpellWord` struct `Analysis` trait, it makes me rethink the approach. Now I see 2 ways:

1. Rust-analyzer takes care of the spell checking in a _**tightly-coupled**_ manner, it uses _libraries_ `typos-dict / typos-vars` and is responsible to find and apply the dictionary configs.
2. In a more _**loosely-coupled**_, Rust-analyzer does the heavy lifting logic to find only the relevant tokens to spell check (_definition_ identifiers...) and send them via a pipe (std in ?) to the _binary_ `typos-cli`, and it replies via the pipe if it's good or not (in json format). Here `typos-cli` is responsible to find and apply the dictionary configs.

The advantage of the _**tightly-coupled**_ is that it might be easier as we will have more control and flexibility. 
The disadvantage is the extra configs files that Rust-analyzer will be responsible.

The advantage of the _**loosely-coupled**_ is that it's a more generic solution, it can be nice to define a simple protocol to send only the relevant token to spellcheck and the way to repond to those request, it could then be used in others LSP (Python, Zig, ...) to provide spell check there too. The spelling configs is not the LSP responsability. Also, if other spell checker implements the same protocol, it would work too.
The disadvantage is that implemented trivially won't be performant to spawn a process for each request, so now we will need to adapt the binaray `typos-cli` to a long live process, with a live watcher for the spelling config files and hot reload, that will listen for spelling request and respond.

#### About updating the config (idea):
For the user, it would be great that in case of spelling error detected, we would have the choice to ether _fix_ it with the suggested spelling (if present) or to _add_ the new word to the spelling config.
In the case of the **tightly-coupled**, this is doable internally because r-a is watching the config files already.
In the case of the **loosely-coupled**, it requires a special request via the pipe to update the config and the typos will update the config.

#### More about the spelling config:
I have no clear idea, only that from my experience and whish list, I would like those settings:

- Extra words that are case insensitive
- Extra words that are case sensitive (such as proper names ...)
- Extra regexes to ignore (emails addresses ...)
- Files to skip (respect .gitignore on or off + list of files and globs)

Typos has already a [config file (typos.toml, _typos.toml, or .typos.toml)](https://github.com/crate-ci/typos/blob/master/docs/reference.md), it's not 100% clear to me all the options but it looks ok so far

> Search parents of specified file / directory for one of typos.toml, _typos.toml, or .typos.toml

This should be extend I think. Imagine a monorepo, with a workspace at the root, I can see a use to have a typos.toml at the root level, with extra words such as the company name for example, and in some of the members crates, you would have an extra typos.toml with some local config specific to the crate (a typo that the user prefer to ignore in this crate only because it's too much effort to fix ?). I would expect the spell checker to apply the root config and to update on it the local config, merging and extending the root config.

#### Clippy:
As @epage pointed out and this open issue https://github.com/rust-lang/rust-clippy/issues/2508, maybe indeed the spell checking could be integrated in the linter instead of the LSP? should it be tightly-coupled or loosely-coupled ? I don't know.
I am tagging @oli-obk and @flip1995 to see what they think of this.

**Reply by flip1995** [reply]
Integrating it in Clippy doesn't quite work: Clippy doesn't see normal comments (`//`, `/* */`), only doc comments (`///`, `//!`). So even when integrating it in Clippy, which already has the information what a keyword/ident/... is, would skip over all comments. And IMO spell checking comments is the most useful thing about spell checking.

It could be integrated into Clippy, if Clippy would re-lex the entire crate, but we don't want to do that for several reasons. But if we were to implement this, then one could also just implement another `cargo` tool that has the same access to `rustc` internals and information, use the `rustc` lint+diagnostic infra and could be called through r-a - just like you can replace `cargo check` with `cargo clippy` in the config, you could replace `cargo check` with `cargo spell`.

EDIT: A quick google search revealed [`cargo-spellcheck`](https://crates.io/crates/cargo-spellcheck), which currently just runs on doc comments IIUC.

**Reply by epage** [reply]
>  And IMO spell checking comments is the most useful thing about spell checking.

For me, its the opposite.  My original motivation for typos was because I had to deal with a typo in a public API and wanted to prevent that from happening again.  Its in all of the outward facing areas (APIs, doc comments, strings) that I most care about fixing typos.

**Reply by GilShoshan94** [reply]
> And IMO spell checking comments is the most useful thing about spell checking.

100% agree

> EDIT: A quick google search revealed [cargo-spellcheck](https://crates.io/crates/cargo-spellcheck), which currently just runs on doc comments IIUC.

cargo-spellcheck is cool too, thank you for sharing, I looked under and it uses [nlprule](https://github.com/bminixhofer/nlprule), which is described as:
>> A fast, low-resource Natural Language Processing and Error Correction library written in Rust. nlprule implements a rule- and lookup-based approach to NLP using resources from [LanguageTool](https://github.com/languagetool-org/languagetool).

And looking more into [languagetool](https://languagetool.org/), it's very cool, especially the rephrase suggestion, feels like chatgpt for writting better comment. 

I'm defining this as out of scope for now for a spell checker.

But I have to admit that having something like this integrated in Rust-Analyzer would be an amazing DX, your code get all the existing LSP features r-a is already providing, and in addition you get spell check without all the noise since it would be only on the relevant places (on the definitions and comments) AND you would get some NLP AI to rephrase comments, all of this in one tool, one interface! 

That's the future one day (I know Git Copilot and ChatGPT exist already but I don't trust it so much to not collect info...)

**Reply by GilShoshan94** [reply]
> For me, its the opposite. My original motivation for typos was because I had to deal with a typo in a public API and wanted to prevent that from happening again. Its in all of the outward facing areas (APIs, doc comments, strings) that I most care about fixing typos.

100% agree too, it's important in both public API and in documentation, comments.

**Reply by GilShoshan94** [reply]
@flip1995 @matklad @Veykril @epage

Hi, back from vacation.
Can we add this on the roadmap for r-a (I know it's not high priority)?
Can we create an issue from this discussion ?

**Reply by Veykril** [reply]
Sure, an issue to track this seems fine

**Reply by GilShoshan94** [reply]
Done, I opened #14597
